### ISSUE TITLE
Remote Server - Could not show image for "x". please check log

### ISSUE THREAD

**Issue by petrgeiger-incieve** [elazarcoh/simply-view-image-for-python-debugging#5]
Hi,

I try to use extension in VS Code with Remote Server, but it fails and VS code pop ups error message.
`could not show image for "x". please check log`
I would like to provide more information, but I can't see much in logs. I only found in Remote Extension Host logs this.

```
[2021-08-18 13:23:59.968] [exthost] [info] ExtensionService#loadCommonJSModule file:///home/geigerp/.vscode-server/extensions/elazarcoh.simply-view-image-for-python-debugging-2.0.1/dist/extension.js
[2021-08-18 13:24:18.438] [exthost] [error] [elazarcoh.simply-view-image-for-python-debugging] provider FAILED
[2021-08-18 13:24:18.438] [exthost] [error] Error: Unable to find thread to evaluate variable reference.
    at b.handleErrorResponse (vscode-file://vscode-app/snap/code/71/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:987:33041)
    at vscode-file://vscode-app/snap/code/71/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:987:32417
[2021-08-18 13:24:19.679] [exthost] [error] [ms-python.python] provider FAILED
[2021-08-18 13:24:19.679] [exthost] [error] Canceled: Canceled
	at w.makeCancelError (/home/geigerp/.vscode-server/extensions/ms-python.python-2021.8.1105858891/out/client/extension.js:9:171435)
	at w.handleFailedRequest (/home/geigerp/.vscode-server/extensions/ms-python.python-2021.8.1105858891/out/client/extension.js:9:171286)
	at /home/geigerp/.vscode-server/extensions/ms-python.python-2021.8.1105858891/out/client/extension.js:9:131628
	at async s.provideHover (/home/geigerp/.vscode-server/bin/379476f0e13988d90fab105c5c19e7abc8b1dea8/out/vs/server/remoteExtensionHostProcess.js:85:164654)
```

Finally, I would like to really let this work, so if you need more info or any other help, please contact me. One last tip, it would be really nice if the extension would work also with pytorch tensors out of the box.

**Comment by elazarcoh** [comment]
Hi,
I'm trying to reproduce. 

First, I want you to verify that the extension generally works:
1. select some image variable in the editor, while in debug mode
2. click F1 on the keyboard
3. use the `View Image (Python Debug)` command
is it opening the image on the side?

Second, 2 questions for you:
1. When you hover with the cursor on some image variable, does it open a light-bulb beside it? if so, is the menu have anything related to the extension ("View Image")?
2. Is the "image watch" view (it should be in the debug container, probably below "breakpoints") has any variables? to check it, set a breakpoint right after declaring some image variable:
   ```python
    x = imread("some/image.png")
    print("break")  # <- put a breakpoint here
    ```

 <br />

Regarding the pytorch tensor, it's actually was implemented a couple of weeks ago.  It requires the `torchvision` package to operate. You can try using it with `F1 -> View Tensor (Python Debug)`

**Comment by elazarcoh** [comment]
I found & fixed various bugs while trying to reproduce the exact same behavior you're describing.
Please try with the latest version (2.0.3) and see if it still doesn't work.

If so, it would be very helpful if you can give a minimal example where it fails:
* is it happening while hovering on some variable?
* is it happening while executing "View Image" from the command palette?
* is it happening from something in the "image watch" view?

Thanks

**Comment by douglasrizzo** [comment]
Hi, I am running my code on a local and got the same problem. It was a pop-up with the message:

> could not show image for "one_image". please check log.

I've attached what seems to be the relevant log below.

- I've done my tests while in debug mode in VS Code, with the debug cursor stopped right after the line in which the variable containing the image is created.
- I get this error while opening a PyTorch tensor of dimensions (3, 1000, 866) (channels, height, width) with RGB values in the range 0 to 255, when the tensor is either in the CPU or GPU.
- I am using version v2.0.3 of the plugin, installed directly from the VS Code Extensions sidebar.
- The error happens both when using the Image Watch (by clicking the "View" button) as well as when clicking the lightbulb button when the cursor sits on top of the variable.
- - The variable containing the image appears in the Image Watch, as I've configured the plugin to do so.

```log
[2021-09-08 17:57:52.005] [exthost] [error] [elazarcoh.simply-view-image-for-python-debugging] provider FAILED
[2021-09-08 17:57:52.005] [exthost] [error] Error: No debugger available, can not send 'scopes'
    at O.customRequest (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:1943:43577)
    at g.$customDebugAdapterRequest (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:1493:7472)
    at d._doInvokeHandler (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:12737)
    at d._invokeHandler (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:12421)
    at d._receiveRequest (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:11083)
    at d._receiveOneMessage (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:9761)
    at vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:7857
    at l.fire (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:69:1712)
    at r.fire (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:85:16477)
    at f._receiveMessage (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:85:21747)
    at vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:85:18625
    at l.fire (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:69:1712)
    at b.acceptChunk (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:85:13838)
    at vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:85:13186
    at Socket.k (vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js:2872:163769)
    at Socket.emit (events.js:315:20)
    at addChunk (internal/streams/readable.js:309:12)
    at readableAddChunk (internal/streams/readable.js:284:9)
    at Socket.Readable.push (internal/streams/readable.js:223:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:188:23)
```

**Comment by elazarcoh** [comment]
Hi, 
Thanks for the information. I can't reproduce on my end.

See the error message in the log? "No debugger available, can not send 'scopes'"
It's basically means there is no connection with the debugger, which is not related to the extension per-se. 

So, to understand the problem better, what I want to know is:

- Whether it's happening on all your debug sessions, or is it sometime works and sometimes not.
- If you go to the next line with the debugger, is it still happening? can you view the tensor?
- If you put the breakpoint one line before you want it, then make a single line step, then try to see the tensor, is it working?
- When you're using the lightbulb, what exactly happens? does it shows `View Tensor (torch.Tensor)`? Or it is failing before that?
- Can you reproduce with a minimal usage example and post it here?
- Which version is VSCode and the Python extension? Is it the latest for both?

I would really appreciate if you can give me as much information as you can.
Thanks

**Comment by petrgeiger-incieve** [comment]
> Hi,
> I'm trying to reproduce.
> 
> First, I want you to verify that the extension generally works:
> 
> 1. select some image variable in the editor, while in debug mode
> 2. click F1 on the keyboard
> 3. use the `View Image (Python Debug)` command
>    is it opening the image on the side?
> 
> Second, 2 questions for you:
> 
> 1. When you hover with the cursor on some image variable, does it open a light-bulb beside it? if so, is the menu has anything related to the extension ("View Image")?
> 2. Is the "image watch" view (it should be in the debug container, probably below "breakpoints") has any variables? to check it, set a breakpoint right after declaring some image variable:
>    ```python
>     x = imread("some/image.png")
>     print("break")  # <- put a breakpoint here
>    ```
> 
> 
> Regarding the pytorch tensor, it's actually was implemented a couple of weeks ago. It requires the `torchvision` package to operate. You can try using it with `F1 -> View Tensor (Python Debug)`

First, sorry for late reply, I somehow missed a notification. I have got a version 2.0.3 and there is my minimal testing code.
```
import torch

t = torch.randn(256, 256).clamp(0, 1)
x = t.numpy()
t # breakpoint here
```

First, in local instance of vscode everything works. But in remote vscode it does not. For all 3 approaches the error message is the same as in initial post, pop up in bottom right corner of vscode. It appears after either clicking light bulb and selecting view image or after clicking view image in image watch menu or using command. For numpy tensors not even a tab with(out) image is displayed. However , for tensor view the tab with empty image preview is shown, but with a message `an error occurred while loading image`. In the tab I see it points to file `/tmp/svifpd/t.png`, but when I check in terminal there is indeed no such file in this location. This time I see nothing interesting in logs. If you have some other ideas how I could help, I keep checking this thread, thanks.

**Comment by elazarcoh** [comment]
>  For numpy tensors not even a tab with(out) image is displayed. 

This is by design. I guess it can be implemented with scitkit-image [montage](https://scikit-image.org/docs/stable/api/skimage.util.html#skimage.util.montage)  (I don't really like to implement making this montage myself, to keep things simple). I'll add it to my backlog.

> However , for tensor view the tab with empty image preview is shown, but with a message an error occurred while loading image. In the tab I see it points to file /tmp/svifpd/t.png, but when I check in terminal there is indeed no such file in this location. 

That's interesting. First, the only way to get to the point of opening the image is if the saving was successful (means without exceptions). this is the code I'm using:
```python
torchvision.utils.save_image(tensor, path)
```
So, the fact the you got the path means that somehow it were successfully saved (in some sense).

So the first question I have is, is the /tmp/svifpd directory exists? If so, are there any files under it? If so, is any of them a valid image? (one you can open with some image view, e.g. vscode is good enough).

Another thing I want to check if using another path for saving will change anything. 
Please try to turn off this setting `svifpd.useTmpPathToSave`. It will cause the extension to save image to the vscode storage instead.

Another possibility, though I don't have a good idea of how to check it, is that for some reason your remote debugger is slow to response for the requests I'm sending for the debugger to save the images. So, possibly, the task is not failing with an error, but still fails because of some sort of timeout. 
This might be a little complicated to check. I'll try to think of some way. Though, to do it I might need to create a custom version of the extension with more logs and more requests to the debugger. Will you be able to install it if I would make it?


This is what currently comes to my mind. 
I'll let you know if I would think of anything else. Possibly I'll make a version with more logs if I'll think it might help.


If you have anything else you think is relevant it might be helpful.

Thanks

**Comment by petrgeiger-incieve** [comment]
> So the first question I have is, is the /tmp/svifpd directory exists? If so, are there any files under it? If so, is any of them a valid image? (one you can open with some image view, e.g. vscode is good enough).

Directory does exists, but is empty. HOWEVER, I found out that the directory is owned by another user (my colleague - he tried the extension some time before also with no success). I tried to add it absolute write permission to the folder, but no image is still created in it. When I removed the folder it is not created again, thus the extension is not even creating a folder. 

> Another thing I want to check if using another path for saving will change anything.
> Please try to turn off this setting `svifpd.useTmpPathToSave`. It will cause the extension to save image to the vscode storage instead.

Ok, this works - partly. For `View image` it successfully opens image in new tab pointing to some file in `$HOME/.vscode-server/...` - probably as expected. However, for `View Tensor` it also points to somewhere in that folder, but the file is again missing.  I've noticed that `View image` works also for pytorch tensors variables, in which case the image is successfully shown.

> Another possibility, though I don't have a good idea of how to check it, is that for some reason your remote debugger is slow to response for the requests I'm sending for the debugger to save the images. So, possibly, the task is not failing with an error, but still fails because of some sort of timeout.
> This might be a little complicated to check. I'll try to think of some way. Though, to do it I might need to create a custom version of the extension with more logs and more requests to the debugger. Will you be able to install it if I would make it?

I skipping this for now as the problem seems to be with `/tmp` folder.

So, to summary with `svidpd.useTmpPathToSave` unchecked I am now able to show images both of numpy and pytorch tensors using `View Image` - thanks. To add little follow up, I just want to note that the pytorch tensors are not shown in `image watch` menu, but I guess it is not just implemented, yet. Also for pytorch tensors the light bulb is not showed, but F1 command works.

**Comment by elazarcoh** [comment]
Hmm I don't know what is it about the tmp folder that doesn't work, but I think the resolution is good enough. Do you concur?
We can farther investigate it if you like. That sounds like something with the permission of `/tmp`. You might want to check it regardless.

Hopefully, the directory is not going to blow up of many images, because I have some mechanism to not keep too many images, though it might not be completely robust. If it would come to that, I can improve this feature. For now I don't see reason to touch it. 


 > However, for View Tensor it also points to somewhere in that folder, but the file is again missing.

To understand it better: when you say "View Tensor it also points to somewhere in that folder", do you mean it open an image file inside `$HOME/.vscode-server/...`, but the tab is left empty? can you make sure it's the same folder as for the `View Image`? Do you have torchvision installed? what happens if you:
```python
import torch

t = torch.randn(256, 256).clamp(0, 1)
torchvision.utils.save_image(t, "some/path/t.png")
```
Do you get an image `t.png` with the tensor? Maybe I have the python implementation wrong at my side.


> I've noticed that View image works also for pytorch tensors variables, in which case the image is successfully shown.

That should not happen 🤔 unless I misunderstood.
```python
import torch

t = torch.randn(256, 256).clamp(0, 1)
x = t.numpy()
t # breakpoint here
```
`View Image` should only work for `x`  (you should not see `View Tensor` in the light-bulb)
`View Tensor` should only work for `t`
do you see a different behavior?


> To add little follow up, I just want to note that the pytorch tensors are not shown in image watch menu, but I guess it is not just implemented, yet. Also for pytorch tensors the light bulb is not showed, but F1 command works.

It actually is! You just need to configure it. see https://github.com/elazarcoh/simply-view-image-for-python-debugging#watch-view.
I intend to have a menu with checklist for it, so it would be more intuitive, but couldn't get to it just yet.

**Comment by petrgeiger-incieve** [comment]
> Hmm I don't know what is it about the tmp folder that doesn't work, but I think the resolution is good enough. Do you concur?
> We can farther investigate it if you like. That sounds like something with the permission of `/tmp`. You might want to check it regardless.

We can leave it. But as I mentioned in previous comment I tried to allow all permisions `rwx rwx rwx` and without effect and if the directory is removed it is no recreated.


> 
> > However, for View Tensor it also points to somewhere in that folder, but the file is again missing.
> 
> To understand it better: when you say "View Tensor it also points to somewhere in that folder", do you mean it open an image file inside `$HOME/.vscode-server/...`, but the tab is left empty? can you make sure it's the same folder as for the `View Image`? Do you have torchvision installed? what happens if you:

Yeah it is the same folder and there is no image produced by `View Tensor` in that folder. `View Image` creates file in the same folder file and it is successfully showdd in vscode tab. And yes I have torchvision installed.
```
~/.vscode-server/data/User/workspaceStorage/37e4c781fa5422cf5f49c898f1f47055/elazarcoh.simply-view-image-for-python-debugging
```

> ```python
> import torch
> 
> t = torch.randn(256, 256).clamp(0, 1)
> torchvision.utils.save_image(t, "some/path/t.png")
> ```
> 
> Do you get an image `t.png` with the tensor? Maybe I have the python implementation wrong at my side.

Yeah, it works.

> 
> > I've noticed that View image works also for pytorch tensors variables, in which case the image is successfully shown.
> 
> That should not happen  unless I misunderstood.
> 
> ```python
> import torch
> 
> t = torch.randn(256, 256).clamp(0, 1)
> x = t.numpy()
> t # breakpoint here
> ```
> 
> `View Image` should only work for `x` (you should not see `View Tensor` in the light-bulb)
> `View Tensor` should only work for `t`
> do you see a different behavior?

The light bulb shows correctly only for `x` as you mention. But when I leave cursor on `t` and press `F1` choose `View Image`, it displays image correctly. Doing `F1` `View Tensors` on `t` fails as mentioned.

**Comment by elazarcoh** [comment]
OK, thanks for the information. I think I figured it out.
For tensors, I assumed it to have at either 3 or 4 dimensions. 
That is:
```python
len(t.shape) == 3 and t.shape[0] in (1, 3)  # gray or rgb
or
len(t.shape) == 4 and t.shape[1] in (1, 3)  # gray or rgb
```
My main issue with changing it is the ability to detect if a tensor is a valid "viewable" object (image/multiple images), and I don't like the idea of using `try: ... except: ...` to detect it.
I'm going to change it to:
```python
def is_torch_tensor(obj):
    valid_channels = (1, 2, 3, 4)
    try:
        is_valid = isinstance(obj, torch.Tensor)
        is_valid &= len(obj.shape) in (2, 3, 4)
        if len(obj.shape) == 2:
            pass
        elif len(obj.shape) == 3:
            is_valid &= obj.shape[0] in valid_channels
        elif len(obj.shape) == 4:
            is_valid &= obj.shape[1] in valid_channels
        return is_valid
    except:
        return False
```
I hope it'll be sufficient to capture all cases.

I'll probably need to change it to support channels last sometime (AFAIK it's not officially supported yet, in any case).

**Comment by elazarcoh** [comment]
Just published new version (2.0.4) addressing the tensor issue.

@douglasrizzo seems to me that your issue is not related. I'm closing this issue. Please open a new issue if you still need help. Thanks
